src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs

```diff
@@ -368,19 +401,23 @@ public static ColInfo Create(string name, PrimitiveDataViewType itemType, Segmen
                 if (segs != null)
                 {
                     var order = Utils.GetIdentityPermutation(segs.Length);
-                    Array.Sort(order, (x, y) => segs[x].Min.CompareTo(segs[y].Min));
 
-                    // Check that the segments are disjoint.
-                    for (int i = 1; i < order.Length; i++)
+                    if (segs[0].Name is null)
```

Comment:
I don't understand this check. Why is `segs[0]` special here?

Also- what happens if `segs` is non-null, but is empty?

Reply:
We don't support mixing index based and name based segments, so if the first segment has a name, all of them must be names.

As for `non-null`, but empty, is that supported; why wouldn't we just normalize that to `null`?

Reply:
> We don't support mixing index based and name based segments, so if the first segment has a name, all of them must be names.

I think that would be a good comment to stick in here. It isn't immediately obvious to readers.

> As for non-null, but empty, is that supported

Walking backward through the code, I see that Column.Source is a publicly settable array. An external caller could set that to an empty array. In the `Bindings` constructor, if Column.Source is an empty array, it would cause `segs` here to be an empty array.

---
In reply to: [332593763](https://github.com/dotnet/machinelearning/pull/4308#discussion_r332593763) [](ancestors = 332593763)

Reply:
Updated to handle an empty segment.